File: frontend/src/types/question.ts
export type PageMeta = {
  page: number;
  page_size: number;
  total: number;
};

export type QuestionListItem = {
  id: number;
  difficulty_level: number | null;
  blank_count: number;
  has_subquestions: boolean;
  source_status: string;
  annotation_status: string;
  required_annotations: number;
  annotation_count: number;
  latest_embedding_version: string | null;
  subject: {
    id: number;
    code: string;
    name: string;
  };
  grade?: {
    id: number;
    grade_index: number;
    grade_name: string;
    edu_stage?: string | null;
  } | null;
  question_type?: {
    id: number;
    code: string;
    name: string;
  } | null;
  content?: {
    stem_text: string;
    stem_html?: string | null;
    answer_text?: string | null;
    solution_text?: string | null;
  } | null;
};

export type QuestionListResponse = {
  items: QuestionListItem[];
  meta: PageMeta;
};

export type QuestionListParams = {
  page?: number;
  page_size?: number;
  keyword?: string;
  subject_id?: number;
  grade_id?: number;
  question_type_id?: number;
  annotation_status?: string;
  source_status?: string;
   question_ids?: number[];
};

export type QuestionExternalRef = {
  id: number;
  external_question_id: string;
  external_type?: string | null;
  is_primary: boolean;
  data_source_code: string;
  data_source_name: string;
};

export type QuestionKnowledgePoint = {
  id: number;
  knowledge_point_id: number;
  knowledge_point_name: string;
  priority: number;
  is_core: boolean;
  is_exam_point: boolean;
  is_last_exam_point: boolean;
};

export type QuestionCatalog = {
  id: number;
  catalog_id: number;
  catalog_name: string;
  school_code?: string | null;
};

export type QuestionDetail = QuestionListItem & {
  external_refs: QuestionExternalRef[];
  knowledge_points: QuestionKnowledgePoint[];
  catalogs: QuestionCatalog[];
};

File: backend/app/schemas/question.py
from pydantic import BaseModel, ConfigDict

from app.schemas.pagination import PageMeta


class SubjectSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    name: str


class GradeSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    grade_index: int
    grade_name: str
    edu_stage: str | None = None


class QuestionTypeSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    name: str


class QuestionContentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    stem_text: str
    stem_html: str | None = None
    answer_text: str | None = None
    solution_text: str | None = None


class ExternalRefOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    external_question_id: str
    external_type: str | None = None
    is_primary: bool
    data_source_code: str
    data_source_name: str


class QuestionKnowledgePointOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    knowledge_point_id: int
    knowledge_point_name: str
    priority: int
    is_core: bool
    is_exam_point: bool
    is_last_exam_point: bool


class QuestionCatalogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    catalog_id: int
    catalog_name: str
    school_code: str | None = None


class QuestionListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    difficulty_level: int | None = None
    blank_count: int
    has_subquestions: bool
    source_status: str
    annotation_status: str
    required_annotations: int
    annotation_count: int
    latest_embedding_version: str | None = None
    subject: SubjectSummary
    grade: GradeSummary | None = None
    question_type: QuestionTypeSummary | None = None
    content: QuestionContentOut | None = None


class QuestionListResponse(BaseModel):
    items: list[QuestionListItem]
    meta: PageMeta


class QuestionDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    difficulty_level: int | None = None
    blank_count: int
    has_subquestions: bool
    source_status: str
    annotation_status: str
    required_annotations: int
    annotation_count: int
    latest_embedding_version: str | None = None
    subject: SubjectSummary
    grade: GradeSummary | None = None
    question_type: QuestionTypeSummary | None = None
    content: QuestionContentOut | None = None
    external_refs: list[ExternalRefOut]
    knowledge_points: list[QuestionKnowledgePointOut]
    catalogs: list[QuestionCatalogOut]
